windows/src/engine/kmtip/tmgrsink.cpp

```diff
@@ -151,21 +151,51 @@ STDAPI CKMTipTextService::OnActivated(REFCLSID clsid, REFGUID guidProfile, BOOL
 {
   LogEnter();
 
-  guidActiveProfile = GUID_NULL;   // I4274
-
-  if(IsEqualGUID(clsid, c_clsidKMTipTextService))
-  {
-    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(c_clsidKMTipTextService, <GUID>, %d)", fActivated);
-    if(fActivated) {
-      guidActiveProfile = guidProfile;
-      TIPNotifyActivate((GUID *)&guidProfile);
+  WCHAR bufClsid[40] = L"", bufProfile[40] = L"";
+  if (StringFromGUID2(clsid, bufClsid, 40) == 0) bufClsid[0] = 0;
+  if (StringFromGUID2(guidProfile, bufProfile, 40) == 0) bufProfile[0] = 0;
+  
+  if(IsEqualGUID(clsid, c_clsidKMTipTextService)) {
+    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(c_clsidKMTipTextService, %s, %d)", bufProfile, fActivated);
+    TIPNotifyActivate(fActivated ? (GUID *)&guidProfile : NULL);
+  } else {
+    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(%s, %s, %d)", bufClsid, bufProfile, fActivated);
+
+    if (fActivated) {
+      BOOL bIsDeactivating = TRUE;
+      ITfInputProcessorProfiles *pInputProcessorProfiles = NULL;
+
+      /* Look up the profile which is being activated, and only deactivate ourselves if it is a
+         Keyboard type TIP.
+      */
+      if (LogSUCCEEDED(CoCreateInstance(CLSID_TF_InputProcessorProfiles, NULL, CLSCTX_INPROC_SERVER,
+        IID_ITfInputProcessorProfiles, (void**)&pInputProcessorProfiles))) {
```

Comment:
Query whether it is safe to call `CoCreateInstance` here or if we should call [`TF_CreateInputProcessorProfiles`](https://docs.microsoft.com/en-us/windows/win32/api/msctf/nf-msctf-tf_createinputprocessorprofiles)

Reply:
As far as I know, this will be okay. We should not be constrained from running COM functions within the auspices of the TIP as it is itself a COM object.